### ISSUE TITLE
Can’t exclude symlink directory from watch task

### ISSUE THREAD

**Issue by timseverien** [gulpjs/gulp#1249]
I have a Gulp watch task that watches for SCSS files:

```js
gulp.task('watch', ['default'], function() {
    gulp.watch(sources.styles, ['styles']);
});
```

When running this task, I get the following error:

`Error: ELOOP, too many symbolic links encountered '/path/to/symlink-directory'`

For the `sources.styles` patterns I’ve tried a handful of things:

```js
// Includes symlink
sources.styles = ['src/styles/**/*.scss'];

// Symlink exclude attempt no. 1
sources.styles = [
  '!src/styles/symlink',
  'src/styles/**/*.scss'
];

// Symlink exclude attempt no. 2
// Based on: https://github.com/gulpjs/gulp/issues/165#issuecomment-32626133
sources.styles = [
  '!src/styles/symlink{,/**}',
  'src/styles/**/*.scss'
];
```

None of the above works. What do?

**Comment by contra** [comment]
@es128 will this be fixed by switching to chokidar?

**Comment by es128** [comment]
I've never seen that error before, but have addressed circular symlinks in the past. Would be easiest for @timseverien to confirm by just setting up a simple chokidar watcher on `sources.styles` and seeing if it throws.

**Comment by jasonmit** [comment]
I am seeing this as well

**Comment by es128** [comment]
@jasonmit perhaps you can help us determine how chokidar fares when watching the same paths that cause the error to be thrown in gulp?

**Comment by timseverien** [comment]
@es128 I’ve tested [chokidar](https://github.com/paulmillr/chokidar) 1.0.5 both with as without the exclude pattern, with the following script:

```js
chokidar.watch([
    // The watcher works both with as without commented line
    //'!src/styles/bootstrap{,**}',
    'src/styles/**/*.scss'
]).on('all', function(event, path) {
    console.log(event, path);
});
```

Both cases worked fine and outputted the following lines:

```
add src/styles/_base.scss
add src/styles/critical.scss
add src/styles/style.scss
add src/styles/components/_footer.scss
add src/styles/helpers/_mixins.scss
```

It appears chokidar ignores symlinks completely.

**Comment by es128** [comment]
> It appears chokidar ignores symlinks completely.

That's not the case generally, but it seems as if you've bumped into https://github.com/paulmillr/chokidar/issues/293 - which occurs because the glob pattern does not match the resolved symlink paths. 

If you set it to just watch all of `src/styles` (no glob) that should expose whether the `ELOOP` problem is still possible, and you should also see events for everything under `bootstrap`.

**Comment by timseverien** [comment]
When I’m not using a glob (`src/styles` or `src/styles/`), the watch task isn’t triggered at all. I just want to watch everything except the bootstrap directory.

**Comment by es128** [comment]
I just meant in the chokidar example. Just trying to determine whether the `ELOOP` error you originally encountered can also happen with chokidar.

**Comment by timseverien** [comment]
Turns out it works if I flip the patterns ¯\_(ツ)_/¯

```js
[
    'src/styles/**/*.scss',
    '!src/styles/bootstrap{,/**}'
]
```

**Comment by phated** [comment]
@timseverien your original problem is chocked up to #651; however, we are trying to get the file watcher switched to chokidar. If you are able to continue testing your setup with chokidar as @es128 mentioned, that would be very helpful.  Thanks.

**Comment by timseverien** [comment]
@phated @es128 Oh, sorry for the misunderstanding. Before flipping the patterns, I’ve thoroughly tested different patterns (below) with chokidar and I was unable to trigger an error of any kind.

```js
// ✓ Exclusively inclusive ;)
'src/styles/**/*.scss'

// ✓ Directory exclusion
[
    '!src/styles/boostrap',
    'src/styles/**/*.scss'
]

// ✓ Exclusion glob
[
    '!src/styles/bootstrap{/,**}',
    'src/styles/**/*.scss'
]
```

Again, all of these worked with chokidar.

**Comment by phated** [comment]
Now that chokidar has been merged into the 4.0 branch. Can we close this?

**Comment by es128** [comment]
afaict, yes

If the `ELOOP` error can still be reproduced, it ought to be reported to chokidar's tracker.
